### PR TITLE
Deploy Docusaurus websites to your forked gh-pages branch

### PR DESCRIPTION
**Pull request by elliot-nelson**
### SUMMARY

Eventually, all of the different website projects will live in the `websites/` folder of this repo, and we want to make it really easy for contributors to the documentation to push up a live version of their own branch to share with others, to test on tablets and phones, etc.

This PR adds the ability for a contributor to fork the repo, clone their fork, and run:

```console
rush install
rush build
GIT_USER=<myself> rush deploy-fork
```

This command will create or overwrite the `gh-pages` branch of the user's fork with copies of all of the supported websites, automatically linked up for cross-site linking between each other, accessible at the following URLs:

```text
https://<myself>.github.io/rushstack-websites/rushjs.io/
https://<myself>.github.io/rushstack-websites/rushstack.io/
https://<myself>.github.io/rushstack-websites/tsdoc.org/
# etc.
```

### DETAILS

Accomplishing the above requires a bit of coordination between the settings used when building and when deploying, which I think in general we want to be separate steps, so we can eventually take advantages of features like build caching.

A new `site-config` project was added to consolidate the decision making about what _target_ the user is building for.  See the README in this PR for more information on "targets".

To accomplish deploying all of the sites in subfolders, a new `deploy-fork` project was added that collects all the other sites as dependencies, and then uses Docusaurus' own deploy command to push the customized `build` folder up to the contributor's fork.

### TESTING

You can see the result of the `deploy-fork` command in my own copy:
 - https://elliot-nelson.github.io/rushstack-websites/rushstack.io/pages/index/

### CODE REVIEW

File: tools/deploy-fork/scripts/build.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+const BUILD_FOLDER = 'build';
+
+process.chdir(`${__dirname}/..`);
+
+FileSystem.deleteFolder(BUILD_FOLDER);
+
+const rushConfig = JsonFile.load(path.join('..', '..', 'rush.json'));
+
+for (let project of rushConfig.projects) {
```
Comment (octogonz) [inline tools/deploy-fork/scripts/build.js:12]: Should we use `@microsoft/rush-lib` for this?  The `projects` array is likely to move out of `rush.json` pretty soon, because it can be huge.

File: tools/site-config/src/index.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// Halp!
```
Comment (octogonz) [inline tools/site-config/src/index.js:1]: ?

File: tools/site-config/src/index.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            console.log(
+                `Detected local dev server, defaulting to TARGET=local.\n` +
+                `(You can override this by specifying TARGET=[local, fork, prod])`
+            );
+            return { target: 'local' };
+        } else {
+            const org = getGitHubOrg();
+            if (org === 'microsoft') {
+                console.log(
+                    `Detected prod build on main repo, defaulting to TARGET=prod.\n` +
```
Comment (octogonz) [inline tools/site-config/src/index.js:51]: How likely is it that this "detection" heuristic might fail?  I probably have permissions to push to the `gh-pages` branch -- I wouldn't want to accidentally push to the wrong target

File: tools/deploy-fork/scripts/build.js (hunk 1 above)
Comment (octogonz) [reply tools/deploy-fork/scripts/build.js:12]: I probably would not mind having to explicitly do `rush deploy-fork --target=prod`.

At least in the past, 94% of the time, testing would be with the localhost dev server, and 4% maybe of the time the CI machine is pushing to production.  And maybe only about 2% of the time did it seem desirable to demo my work in progress for another person, who can't easily run the dev server themselves.  But maybe that will be changing... :-)

**Review by octogonz — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -5,6 +5,58 @@
...
+
+You can also force this target by setting the environment variable `TARGET=local`.
+
+### "fork"
+
+The `fork` target is intended for use when you are building static Docusaurus sites for deployment,
+but you'll be deploying them to GitHub Pages on your fork of the `rushstack-websites` project. This is
+useful for deploying demo sites to share with others, to test on phones and tablets, etc. This target
+will automatically be selected if you run a production build (`rushx build`) of a website project and
+you have cloned a _fork_ of the `rushstack-websites` project.
```
Comment (octogonz) [inline README.md:26]: I wonder if search engines might accidentally index these sites, if you imagine tons of old deployments sitting around with realistic-looking URLs like `https://elliot-nelson.github.io/rushstack.io-website/` and also hyperlinks pointing to them from real GitHub discussions.

Maybe we could add some HTML metadata to prevent them from getting crawled.

File: tools/site-config/src/index.js (hunk 3 above)
Comment (elliot-nelson) [reply tools/site-config/src/index.js:51]: It depends what you mean by "wrong".

Let's say, for example, you were to clone the https:// rushstack-website project locally by accident, and you happen to have push privileges to a whole bunch of repos.

Building (rush build) anything in your local clone will assume you want `TARGET=prod`, which is correct -- deploying from that clone is intended to push to a website.

Attempting to run `rush deploy-fork` would generate an error, because `microsoft/rushstack-websites` is not supposed to have its own gh-pages branch.

If you _wanted_ to actually deploy to production, you would `cd` into an individual website project and run `rushx deploy`.  For a normal person, that would generate access errors; for you, that might work, but that would be OK.

TLDR: `rushx deploy` in individual websites should only be done with TARGET=prod, and won't work for most contributors.  `rush deploy-fork` is for deploying ALL websites as a package to a fork repo, and will work for anyone that is cloning a fork.
